### PR TITLE
‼️ REFACTOR: HTML and CSS restructuring

### PR DESCRIPTION
**Pull request by choldgraf**
This is a refactor of our HTML and CSS, with the goal of improving a number of the UI/UX issues on the book theme, tightening up our white space and sizing, using a more sensible set of CSS classes that will make the theme behave more predictably, and adding more functionality for the in-page TOC on mobile widths.

This doesn't remove any functionality, though it will likely break any custom CSS rules that were highly dependent on a particular HTML layout.

Major improvements

- We now use `sticky-top` to place our sidebar and topbar, rather than `fixed`, which means it will flow more naturally if things like banners are placed at the top of the page.
- Sidebar/topbar/toc font is slightly smaller than the main text font to create more space
- There's more space for the sidebar + toc bar
- Increased padding for the main content, and improved padding behavior on wide screens + with sidebar collapsed
- Programmatic definition of our margin size and placement, so that there's less hacky manually-chosen numbers in there
- The right sidebar is now viewable on mobile, via a button to toggle the toc similarly to the left sidebar
- The topbar buttons are now divided into two div groups (left and right) and use flexbox to position and center them nicely
- Re-arranges and modularizes our SCSS structure to follow more best practices
- Re-arranges our HTML template structure to match more closely the sphinx-basic-ng theme
- Adds `prettier` pre-commit hook for our javascript and scss files (we already use `black` for Python files so this follows the same convention)

Note - this also bumps our pydata-sphinx-theme back **down** to the 0.7 series. I think 0.8 brought a new version of bootstrap (4.6), and will cause some unexpected behavior, so let's bump that version in an explicit PR.

closes https://github.com/executablebooks/sphinx-book-theme/issues/392
closes https://github.com/executablebooks/sphinx-book-theme/issues/459 (I think?)
checks off the last two boxes in https://github.com/executablebooks/sphinx-book-theme/issues/416
supercedes https://github.com/executablebooks/sphinx-book-theme/pull/471

### Todo

- [x] Cleanup
    - [x] See if there is any other low-hanging fruit to make this theme's structure closer to sphinx-basic-ng
    - [x] Pull out some component-specific SCSS rules into files in the `components/` folder
	- [x] standardize the behavior of buttons with a single css class instead of using one-offs
	- [x] Fix print formatting
	- [x] Topbar buttons should be bigger on mobile, and their coloring seems incorrect on mobile. Maybe stop using the bootstrap `btn` class
- [x] Docs
	- [x] add the use of banner and topbar sections
	- [x] Add an "architecture.md" file and/or otherwise explain the layout of the theme
	- [x] Take a pass at developer docs
          - https://github.com/executablebooks/sphinx-book-theme/pull/469#discussion_r791468960
- [x] QA
	- [x] Look at all our reference pages and make sure there are no regressions on mobile/desktop
	- [x] Make sure that we didn't miss any CSS breakages with the latest docutils
	- [x] Tests are passing
	- [x] Make sure the diffs in regression tests don't seem like we've introduced a bug

### Demos

Sidebar behavior 🎉 

Slack_tohCGXgFs4

cc @jacobtomlinson as this will likely accept the Dask theme since it rearranges a bit of HTML (but will hopefully make it easier to sub-theme!)

### CODE REVIEW

**Comment by jacobtomlinson** [comment]
Thanks for the ping here @choldgraf, these changes look great and as you say should make our lives easier. I've pinned the Dask theme to `0.2.0` for now as this change does break some things for us, but once this is merged and released I'll port our theme over.

**Comment by AakashGfude** [comment]
Thanks, @choldgraf for the improvements here and for creating folder structures for the scss code. While we are restructuring scss, I have an opinion about the architecture of scss files being adopted. Personally, I believe, we should have separate stand-alone files for site-wide variables, mixins, and functions, and name the files as such. An architecture pattern outlined in https://sass-guidelin.es/#architecture was one of the inspirations. A boilerplate GitHub repo for the same https://github.com/KittyGiraudel/sass-boilerplate/tree/master/stylesheets/abstracts. What do you think?

**Comment by choldgraf** [comment]
@AakashGfude that's a great resource, thanks for sharing. I'll re-work the SCSS to follow this structure, I think it's great to have some kind of objective standard that we shoot for 👍

**Comment by choldgraf** [comment]
@jacobtomlinson here are some docs on how custom header HTML can be added to the site: https://sphinx-book-theme--472.org.readthedocs.build/en/472/customize/header.html

maybe that's something you could re-use?

File: src/sphinx_book_theme/assets/styles/base/_base.scss [hunk 1]
```diff
@@ -2,6 +2,15 @@
 /*********************************************
 * Utility classes used in a few places
 *********************************************/
+// For the helper pixel that we can watch to decide whether we'vez
```
Comment (pradyunsg) [inline src/sphinx_book_theme/assets/styles/base/_base.scss:5]: we'vez?

Comment (choldgraf) [reply src/sphinx_book_theme/assets/styles/base/_base.scss:5]: We'vez is the new warez 😄

Good catch lol

File: docs/customize/header.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  ```console
+  $ cd docs/
+  $ mkdir _templates
+  ```
+- Add the `_templates/` folder to your Sphinx configuration templates:
+
+  ```python
+  templates_path = ["_templates"]
+  ```
+- Create the following template file and add HTML to id:
```
Comment (AakashGfude) [inline docs/customize/header.md:20]: should be "to it"?

**Comment by choldgraf** [comment]
OK tests are passing! It would be great if others would give feedback and give look at the changes. This touches a lot of different parts of the package (thought doesn't fundamentally change its functionality). Can anybody spot some more low-hanging fruit?

**Comment by choldgraf** [comment]
### Update from meeting

We discussed this PR in our EBP meeting today and people generally thought that this was a nice improvement. I'll leave it open a bit longer in case others have any thoughts or suggestions they'd like to make!

**Comment by choldgraf** [comment]
I'm going to plan on merging this one in later today, unless somebody suggests we should hold off. We can continue iterating on the structure, but since this one touches so many parts of the project I think it'll be easier for us to wrap our heads around things if we can do it in more iterative steps.

**Comment by mathbunnyru** [comment]
@choldgraf let's merge this? :)

**Comment by choldgraf** [comment]
Will try to get to this today - ended up with a sick toddler at home so wasn't as productive as I hoped yesterday 😬

**Comment by mathbunnyru** [comment]
Oh, sorry for bothering you.
I wish your toddler to get well!

**Comment by choldgraf** [comment]
OK I made a few last-second docs updates and minor updates to CSS. I think there are still some things to figure out with the extensions integration, but figured it'd be better to merge this in so that we can iterate on those issues in smaller PRs that are easier to review. I opened up https://github.com/executablebooks/sphinx-book-theme/issues/476 to track that ongoing work.

This is a big change so we'll probably make a pre-release eventually, but it'd be great if in the meantime folks tried out `master` (cc @jacobtomlinson and maybe @mathbunnyru would be willing to as well?)

**Comment by mathbunnyru** [comment]
Anyone who interested to try this out, if you use requirements.txt, just add `git+https://github.com/executablebooks/sphinx-book-theme@master#egg=sphinx-book-theme`.

I've tried this and it looks great!
https://jupyter-docker-stacks--1602.org.readthedocs.build/en/1602/using/specifics.html

The feature from docutils 17.0 works well!

**Pull request merged**